File: src/sam/http/api/locations.rs
// ███████     █████     ███    ███
// ██         ██   ██    ████  ████
// ███████    ███████    ██ ████ ██
//      ██    ██   ██    ██  ██  ██
// ███████ ██ ██   ██ ██ ██      ██ ██
// Copyright 2021-2026 The Open Sam Foundation (OSF)
// Developed by Caleb Mitchell Smith (ktheindifferent, PixelCoda, p0indexter)
// Licensed under GPLv3....see LICENSE file.

use rouille::post_input;
use rouille::Request;
use rouille::Response;

pub fn handle(
    _current_session: crate::sam::memory::cache::WebSessions,
    request: &Request,
) -> Result<Response, crate::sam::http::Error> {
    if request.url() == "/api/locations" {
        let objects = crate::sam::memory::Location::select(None, None, None, None)?;
        return Ok(Response::json(&objects));
    }

    if request.url().contains("/api/locations") && request.url().contains("/rooms") {
        let url = request.url().clone();
        let split = url.split("/");
        let vec = split.collect::<Vec<&str>>();
        let location_oid = vec[3];

        if request.method() == "GET" {
            let mut pg_query = crate::sam::memory::PostgresQueries::default();
            pg_query
                .queries
                .push(crate::sam::memory::PGCol::String(location_oid.to_string()));
            pg_query.query_columns.push("location_oid =".to_string());

            let rooms = crate::sam::memory::Room::select(None, None, None, Some(pg_query))?;

            return Ok(Response::json(&rooms));
        }

        if request.method() == "POST" {
            let input = post_input!(request, {
                name: String
            })?;

            let mut room = crate::sam::memory::Room::new();
            room.name = input.name;
            room.location_oid = location_oid.to_string();
            room.save().unwrap();

            let mut pg_query = crate::sam::memory::PostgresQueries::default();
            pg_query
                .queries
                .push(crate::sam::memory::PGCol::String(room.oid.clone()));
            pg_query.query_columns.push("oid =".to_string());

            let objects = crate::sam::memory::Room::select(None, None, None, Some(pg_query))?;
            if !objects.is_empty() {
                if request.url().contains(".json") {
                    return Ok(Response::json(&objects[0]));
                } else {
                    let response = Response::redirect_302("/locations.html");
                    return Ok(response);
                }
            } else {
                return Ok(Response::empty_404());
            }
        }
    }

    Ok(Response::empty_404())
}

File: www/assets/js/locations.js
// ███████     █████     ███    ███    
// ██         ██   ██    ████  ████    
// ███████    ███████    ██ ████ ██    
//      ██    ██   ██    ██  ██  ██    
// ███████ ██ ██   ██ ██ ██      ██ ██ 
// Copyright 2021-2026 The Open Sam Foundation (OSF)
// Developed by Caleb Mitchell Smith (ktheindifferent, PixelCoda, p0indexter)
// Licensed under GPLv3....see LICENSE file.

$(document).ready(function() {

    
    $.get("/api/locations", function( location_data ) {

       
        $(location_data).each(function() {
            var location = this;


          



           
            var html = "";
            var rooms_html = "";
            $.get(`/api/locations/${location.oid}/rooms`, function( room_data ) {
  

                rooms_html = `
            
                <div class="modal fade" id="manage_rooms_for_location_${location.oid}_modal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true">
                    <div class="modal-dialog" role="document">
                    <div class="modal-content">
                        <div class="modal-header">
                        <h5 class="modal-title" id="exampleModalLabel">${location.name}</h5>
                        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
                            <span aria-hidden="true">&times;</span>
                        </button>
                        </div>
                   
                        <div class="modal-body">`;

                        if(room_data.length == 0){
                            rooms_html += 'No rooms yet...'
                        } else {

                            rooms_html += `
                            
                            <table class="table tablesorter">
                                <thead class=" text-primary">
                                    <tr>
                                        <th>
                                            Name
                                        </th>
                                        <th class="text-center">
                                            
                                        </th>
                                    </tr>
                                </thead>
                                <tbody>`;

                            
                            

                            $(room_data).each(function() {
                     
                                rooms_html += `<tr>
                                    <td>${this.name}</td>
                                </tr>`;


                            });

                            rooms_html += `</tbody>
                            </table>`;
                        }
                        
  
                    rooms_html += `</div>
                        <div class="modal-footer">
                            <button type="button" onclick="addRoomToLocation('${location.oid}');" class="btn btn-primary float-right">Add Room</button>
                        </div>
                        
                    </div>
                    </div>
                </div>
                
                `;

                html += `
                    <div class="col-lg-6 col-sm-12" id="location_${location.oid}">
                        <div class="card ">
                            <div class="card-header">
                                <h4 class="card-title" style="text-align: center;"><i class="fas fa-lightbulb float-left"></i><span>${location.name}</span></h4>
                            </div>
                            <div class="card-body" id="spotify_card_body">
                                <p>${location.address}</p>
                                <button type="button" data-toggle="modal" data-target="#manage_rooms_for_location_${location.oid}_modal" class="btn btn-primary"><i class="fas fa-map-signs"></i> Manage Rooms</button>
                            </div>
                        </div>
                    </div>
                `;
                html += rooms_html;
                
      
                $("#locations_table_body").append(html);

            });

            
        });
        
    });
    
    
});


function addRoomToLocation(location_oid){

    Swal.fire({
        title: 'New Room',
        input: 'text',
        inputLabel: 'Name',
        showCancelButton: true,
        inputValidator: (value) => {
          if (!value) {
            return 'You need to write something!'
          }
        },
        preConfirm:function(){
            in1= $('#swal2-input').val();


            $.post(`/api/locations/${location_oid}/rooms`, { name: in1 } );


            console.log(in1) // use user input value freely
        }
      })

}